### ISSUE TITLE
Add proper support for class decorators

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#143]
Similar to function decorators implemented in https://github.com/astral-sh/ruff/pull/17017, we need to support class decorators properly.

**Comment by AlexWaygood** [comment]
Mypy's actually on version 1.15.0, and still doesn't properly support class decorators, FWIW! https://mypy-play.net/?mypy=latest&python=3.12&gist=5b5798e2bce1a8a90af1e777d4305294

I agree that we definitely should add support for these at some point (ideally before red-knot becomes generally available), but I'm not sure that full support for class decorators _needs_ to be a before-alpha issue or even a before-beta issue. They're _much_ less important than function/method decorators.

(We'll obviously need to add support for specific decorators such as `@typing.final`, `@typing.runtime_checkable`, `@dataclasses.dataclass`, `typing.dataclass_transform`, though.)

**Comment by mwaskom** [comment]
Just mentioning that we at @modal-labs are watching this issue as decorator wrapping (including of classes) is quite central to the design of our Python API and we'd love have `ty` work out of the box for our users. I added some specific comments about a pattern that we are hoping to see working in this issue: https://github.com/astral-sh/ty/issues/2379

**Comment by BarrensZeppelin** [comment]
@AlexWaygood When you wrote
> But I would not use the return type of the decorator as the class's type in this first PR. I would defer that to a followup change, as I expect this to be much more disruptive and harder to pull off.

Did you mean that it's more disruptive for the ecosystem results (i.e. `mypy-primer`), or that it is hard to implement in `ty`?

If it's the former, would you consider adding a configuration option that opts-in to the "correct" behaviour?
I also have a code base that I cannot migrate from `pyright` due to this issue. 🙂

**Comment by AlexWaygood** [comment]
@BarrensZeppelin thanks for letting us know that this impacts you! That's helpful for us in determining how to prioritise this issue.

We certainly intend to add proper support for class decorators at some point, that's why this issue is still open. Once we have that support, there won't be any reason to have a config option. My comment there was just advising Charlie to make an easy incremental change in one specific PR rather than trying to do everything (including the harder changes) all in one PR.

A config option wouldn't really help if adding support for this feature led to ty crashing all the time, which is what we were seeing in the very early version of Charlie's PR at the time and what prompted that comment from me. If that was the effect of the change, we would not be able to land the change with or without the config option.

That doesn't necessarily mean that the crashes are hard to fix, though — AFAIK we haven't looked too hard at what it would take to implement this feature. Again, I was really just advising an incremental approach in Charlie's work.

**Comment by bartfeenstra** [comment]
In the hopes that this is helpful to understanding potential use cases for this feature request: I would like to see this land as well. I make heavy use of class decorators to attach information to classes, for example to mark a class as a certain kind of plugin for my framework. As these classes must expose this information (e.g. `MyClass.plugin()` returns this plugin information), I currently require all of them to subclass a base class that provides the `plugin()` method for returning the information from the class's decorator.

However,  would like to drop the requirement for such classes to have to subclass anything, as we may not own all classes we want to decorate (they may be from other packages). Therefore, I want to be able to use these decorators to type their return values that when decorating `type[T]`, `type[Intersection[T, DecoratedPluginInformation]]` is returned, where `DecoratedPluginInformation` provides the functionality we now require the plugin base class for. The decorator can alter the class internally just fine, what is lacking at the moment is just the typing support.

**Comment by brendancooley** [comment]
Here's another pattern which I think is similar to @bartfeenstra 's. `Intersection` types are a huge unlock here! But need this to land to get it to work end to end.

```python
"""Minimal example: class decorator that adds an attribute to the decorated class."""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Protocol, TypeVar

if TYPE_CHECKING:
    from ty_extensions import Intersection

# What we want to express: the decorated class gains a .resource attribute
class ResourceEnabled(Protocol):
    """Protocol documenting the contract added by the decorator."""
    resource: Resource

class Resource:
    """The object attached to decorated classes."""

    def __init__(self, name: str, cls: type) -> None:
        self.name = name
        self.cls = cls

    def fetch(self) -> str:
        return f"Fetching {self.name}"

SchemaT = TypeVar("SchemaT")

class Registry:
    """Registry that provides class decorators."""

    def __init__(self, namespace: str) -> None:
        self.namespace = namespace

    def register(
        self,
        name: str | None = None,
    ) -> Callable[[type[SchemaT]], Intersection[type[SchemaT], ResourceEnabled]]:
        """Decorator that adds .resource attribute to a class.

        Returns
        -------
        Callable[[type[SchemaT]], Intersection[type[SchemaT], ResourceEnabled]]
            Decorator that enhances class with .resource attribute.
        """

        def decorator(
            cls: type[SchemaT],
        ) -> Intersection[type[SchemaT], ResourceEnabled]:
            resource_name = name or cls.__name__.lower()

            # Add .resource attribute to the class
            cls.resource = Resource(resource_name, cls)  # type: ignore[attr-defined]

            # Return the enhanced class
            return cls

        return decorator

# Usage
registry = Registry("my_project")

@registry.register(name="users")
class UserSchema:
    id: int
    name: str

# Runtime: works correctly
print(UserSchema.resource.fetch())  # "Fetching users"

# but ty complains
# error[unresolved-attribute]: Class `UserSchema` has no attribute `resource`
```

**Comment by jvdillon** [comment]
Hello! I would like to second my support for this feature.

My project, [configgle🤭](https://github.com/jvdillon/configgle?tab=readme-ov-file#type-safe-make), cannot annotate the existence (using `@autofig`) of the class level attribute to track the nested class parent without this feature. Fwiw, note that `basedpyright` does support this feature.
